File: Server/models/TestResult.js
const mongoose = require('mongoose');

const UserAnswerSchema = new mongoose.Schema({
    question: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'Question',
        required: true
    },
    selectedAnswers: [{
        type: mongoose.Schema.Types.ObjectId,
        ref: 'Question.answers'
    }],
    freeTextAnswer: {
        type: String,
        default: ''
    },
    isCorrect: {
        type: Boolean,
        required: true
    },
    timedOut: {
        type: Boolean,
        default: false
    }
});

const TestResultSchema = new mongoose.Schema({
    user: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'User',
        required: true
    },
    completionTime: {
        type: Date,
        default: Date.now
    },
    userAnswers: [UserAnswerSchema],
    createdAt: {
        type: Date,
        default: Date.now
    }
});

TestResultSchema.virtual('totalQuestions').get(function () {
    return this.userAnswers.length;
});

TestResultSchema.virtual('correctAnswers').get(function () {
    return this.userAnswers.filter(answer => answer.isCorrect).length;
});

TestResultSchema.virtual('score').get(function () {
    if (this.userAnswers.length === 0) return 0;
    return (this.correctAnswers / this.totalQuestions) * 100;
});

TestResultSchema.set('toJSON', { 
    virtuals: true,
    transform: function(doc, ret) {
        ret.score = parseFloat(ret.score.toFixed(1));
        return ret;
    }
});

TestResultSchema.set('toObject', { virtuals: true });

TestResultSchema.virtual('timedOutQuestions').get(function () {
    return this.userAnswers.filter(answer => answer.timedOut).length;
});

module.exports = mongoose.model('TestResult', TestResultSchema);

File: Client/Models/TestResult.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace TestingApp.Models
{
    public class TestResult
    {
        public User User { get; set; }
        public DateTime CompletionTime { get; set; }
        public List<UserAnswer> UserAnswers { get; set; } = new List<UserAnswer>();
        public int TotalQuestions => UserAnswers.Count;
        public int CorrectAnswers => UserAnswers.Count(a => a.IsCorrect);
        public double Score => (double)CorrectAnswers / TotalQuestions * 100;
    }

    public class UserAnswer
    {
        public Question Question { get; set; }
        public List<Answer> SelectedAnswers { get; set; } = new List<Answer>();
        public string FreeTextAnswer { get; set; }
        public bool IsCorrect { get; set; }
        public bool TimedOut { get; set; }
    }
}